062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
+* The other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quotas plugin - will be configured as `type: strimzi`.
+The API will allow to configure the storage limits at which the plugin will start limiting the producers in order to prevent the disk from getting full.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: strimzi
+      produceQuota: 1000000
+      fetchQuota: 1000000
+      minAvailableBytesPerVolume: 5368709120
+      excludedPrincipals:
+        - my-special-user
+    # ... 
+  # ...
+```
+
+The `type: strimzi` quotas API will support the following fields:
+* `produceQuota` for setting the default produce quota
+* `fetchQuota` for setting the default fetch quota
+* `minAvailableBytesPerVolume` for setting the minimal required free space per volume as number of bytes (this field can be configured only when `minAvailableRatioPerVolume`is not set)
+* `minAvailableRatioPerVolume` for setting the minimal required free space per volume as percentage of the overall volume size (this field can be configured only when `minAvailableBytesPerVolume`is not set)
+* `excludedPrincipals` for setting s list of users to be excluded from the applied quotas
+
+Other options might be configured directly by the user in `.spec.kafka.config` section of the `Kafka` CR.
+
+Based on this configuration, the operator will:
+* Set the `client.quota.callback.class` field in Kafka configuration to `io.strimzi.kafka.quotas.StaticQuotaCallback`
+* Set the configuration on how to connect to the Kafka broker as a client using the `client.quota.callback.static.kafka.admin.` options
+* Configures the default produce and fetch quota (`client.quota.callback.static.produce` and `client.quota.callback.static.fetch`)
+* Configure the minimal available disk space per-volume using the `client.quota.callback.static.storage.per.volume.limit.min.available.bytes` and `client.quota.callback.static.storage.per.volume.limit.min.available.ratio` options
+* Set the excluded users for that the quotas will not be applied using the `client.quota.callback.static.excluded.principal.name.list` option
+
+The quota plugin configuration will not be set a forbidden configuration.
```

Comment:
How does the operator know what the rules governing what is acceptable configuration really are? To me that requires an interaction with the plug-in it self which I don't see how the operator can do.

Reply:
Again, I don't follow this. The plugin runs inside Kafka. There is no interaction between the operator and the plugin. And I do not follow what logic are you talking about here. Again, the README of the quotas plugin does not seem to mention anything like that.

Reply:
I interpreted this line as saying the Operator **will not** present invalid configuration to the Kafka Broker (and by extension the plug-in). Thus the operator would need to understand what does and doesn't represent valid configuration for the plug-in. Which is simple enough for something like the callback class but seems difficult to do for mutually exclusive properties. Thus the operator would need some mechanism to understand what the plug-in validation.

Reply:
Well, I would definitely not say `will not present invalid configuration to the Kafka Broker` in bold. It can do some validation. It often cannot do everything and it often relies on the users not being complete idiots (that assumption is a sometimes pushing it a lot, but not easy to avoid it).

It can do some basic validation such as that `minAvailableBytesPerVolume` and `minAvailableRatioPerVolume` should not be set at the same time. But it will probably not have a complex understanding of the whole configuration. So it is definitely not doing everything. 

I think that normally works reasonably well. But if you are aware of any things we should validate, please share them and I can consider them in the proposal.

> Which is simple enough for something like the callback class but seems difficult to do for mutually exclusive properties. Thus the operator would need some mechanism to understand what the plug-in validation.

So, are you saying the plugin has a validation callback I can just call, pass it the 6 or ho many string values and it tells me it is valid?

Reply:
The plug-in does that validation during startup but doesn't currently expose that in any other way. We could expose a validation API but I don't think its really worth the effort for a single plug-in.

Working on the assumption the project only wants to support the plug-ins it packages then I think encoding the validation in the operator is fine.